src/structures/ThreadChannel.js

```diff
@@ -0,0 +1,393 @@
+
+  /**
+   * Adds a member to this thread.
+   * @param {UserResolvable} member The member to add
+   * @param {string} [reason] The reason for adding this member
+   * @returns {Promise<Snowflake>}
+   */
+  addMember(member, reason) {
+    const id = member === '@me' ? member : this.client.users.resolveID(member);
+    if (!id) return Promise.reject(new TypeError('INVALID_TYPE', 'member', 'UserResolvable'));
+    return this.client.api
+      .channels(this.id, 'thread-members', id)
+      .put({ reason })
+      .then(() => id);
+  }
+
+  /**
+   * Makes the client user join the thread.
+   * @returns {Promise<ThreadChannel>}
+   */
+  join() {
+    return this.addMember('@me').then(() => this);
+  }
+
+  /**
+   * Makes the client user leave the thread.
+   * @returns {Promise<ThreadChannel>}
+   */
+  leave() {
+    return this.members.remove('@me').then(() => this);
+  }
+
+  /**
+   * Gets the overall set of permissions for a member or role in this threads' parent, taking into account overwrites.
+   * @param {GuildMemberResolvable|RoleResolvable} memberOrRole The member or role to obtain the overall permissions for
+   * @returns {?Readonly<Permissions>}
+   */
+  permissionsFor(memberOrRole) {
+    return this.parent?.permissionsFor(memberOrRole);
+  }
+
+  /**
+   * Edits the thread.
+   * @param {Object} data The new data for the thread
+   * @param {string} [data.name] The new name for the trhead
+   * @param {boolean} [data.archived] Whether the thread is archived
+   * @param {number} [data.autoArchiveDuration] How long in minutes before the thread is automatically archived,
+   * one of `60`, `1440`, `4320`, or `10080`
+   * @param {number} [data.rateLimitPerUser] The ratelimit per user for the thread in seconds
+   * @param {boolean} [data.locked] Whether the thread is locked
+   * @param {string} [reason] Reason for editing this thread
+   * @returns {Promise<ThreadChannel>}
+   * @example
+   * // Edit a thread
+   * thread.edit({ name: 'new-thread' })
+   *   .then(console.log)
+   *   .catch(console.error);
+   */
+  async edit(data, reason) {
+    const newData = await this.client.api.channels(this.id).patch({
+      data: {
+        name: (data.name || this.name).trim(),
+        archived: data.archived,
+        auto_archive_duration: data.autoArchiveDuration,
+        rate_limit_per_user: data.rateLimitPerUser,
+        locked: data.locked,
+      },
+      reason,
+    });
+
+    return this.client.actions.ChannelUpdate.handle(newData).updated;
+  }
+
+  /**
+   * Sets whether the thread is archived.
+   * @param {boolean} archived Whether the thread is archived
+   * @param {string} [reason] Reason for archiving or unarchiving
+   * @returns {Promise<ThreadChannel>}
+   * @example
+   * // Set the thread to archived
+   * thread.setArchived(true)
+   *   .then(newThread => console.log(`Thread is now ${newThread.archived ? 'archived' : 'active'}`))
+   *   .catch(console.error);
+   */
+  setArchived(archived, reason) {
```

Comment:
Shouldn't this default to true?
Just opening this for discussion.

Reply:
Hmm, I do kind of like that, there are a couple other places where it would have been a similar decision that don't currently. I.e. `VoiceState#setMute`.

I do think this differs slightly from that though, especially since the likelihood of unarchiving this way seems quite low anyways.

Reply:
Yes, you are correct. But I would rather change `VoiceState#setMute` too.

Reply:
I'm not opposed to having `VoiceState#setMute`, `VoiceState#setDeaf` etc. default to `true`.

Reply:
I'm going to do these ones here because they are new, the voicestate ones should be a separate PR